docs/developer/design/authentication.rst

```diff
@@ -0,0 +1,210 @@
+  LDAP, SAML and others.
+* Dex can be seamlessly deployed in a Kubernetes cluster.
+* Dex provides access to a highly customizable UI which is a step closer to
+  good user experience which we advocate for.
+* Dex can act as a fall back Identity Provider incases were the external
+  providers become unable or are not configured.
+
+Rejected design choices
+^^^^^^^^^^^^^^^^^^^^^^^
+
+Static password file Vs OpenID Connect
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Using static password files involves adding new users by updating a static file
+located on every control-plane Node. This method requires restarting the
+Kubernetes API server for every new change introduced.
+
+This was rejected since it is inflexible to operate, requires storing user
+credentials and there is no support for a pluggable external identity providers
+such as LDAP.
+
+X.509 certificates Vs OpenID Connect
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Here, each user owns a signed certificate that is validated by the Kubernetes
+API server.
+
+This approach is not user-friendly that is each certificate has to be manually
+signed. Providing certificates for accessing the MetalK8s UI needs
+much more efforts since these certificates are browser incompatible.
+Using certificates is tedious since the certificate revocation process is also
+cumbersome.
+
+Keycloak Vs Dex
+~~~~~~~~~~~~~~~
+
+Both systems use OpenID Connect(OIDC) to authenticate a user using a
+standard OAuth2 flow.
+
+They both offer the ability to have short lived sessions so that user access
+can be rotated with minimum efforts.
+
+Finally, they both provide a means for identity management to be handled by an
+external service such as LDAP, Active Directory, SAML and others.
+
+Why not `Keycloak <https://www.keycoak.org/>`_?
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Keycloak while offering similar features as Dex and even much more was rejected
+for the following reasons:
+
+- Keycloak is complex to operate(requires its own standalone database) and
+  manage(frequent database backups are required).
+
+- Currently, no use case exist for implementing a sophisticated Identity
+  Provider like Keycloak when the minimal Identity Provider from Dex is
+  sufficient.
+
+Note that, Keycloak is considered a future fallback Identity Provider if
+the need ever arises from a customer standpoint.
+
+Unexploited choices
+~~~~~~~~~~~~~~~~~~~
+
+* `Guard <https://github.com/appscode/guard/>`_
+
+A Kubernetes webhook authentication server by AppsCode, allowing you to
+log into your Kubernetes cluster by using various identity providers such as
+LDAP.
+
+* `ORY Hydra <https://github.com/ory/hydra/>`_
+
+It's an OpenID Connect provider optimized for low resource consumption.
+ORY Hydra is not an identity provider but it is able to connect to existing
+identity providers.
+
+Implementation Details
+----------------------
+
+Implementation steps(High level)
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Iteration 1
```

Comment:
I don't see anything about pre-provisioned admin user and the ability to enforce password change at first login ... Is it something that should appear in another iteration? or maybe we don't have a solution for now?

Reply:
@thomasdanan I think having a pre-provisioned admin user can be accomplished at the ConfigMap level during Dex deployment using the Helm charts.

Having a password change policy on first login would require mostly frontend work and interactions with the Dex gRPC API.

I suggest adding this to another iteration at least the part of enforcing a password change upon first login.
@gdemonet whats your take on this?